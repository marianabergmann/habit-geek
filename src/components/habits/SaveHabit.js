import React, { useState } from "react";

const SaveHabit = ({
  id,
  name: defaultName = "",
  type: defaultType = "",
  frequency: defaultFrequency = "",
  saveHabit,
  deleteHabit,
  goBack
}) => {
  const [name, setName] = useState(defaultName);
  const [type, setType] = useState(defaultType);
  const [frequency, setFrequency] = useState(defaultFrequency);
  const [validityChecked, setValidityChecked] = useState(false);

  const saveAndClose = event => {
    event.preventDefault();
    setValidityChecked(true);

    if (event.target.checkValidity()) {
      saveHabit(id, name, type, frequency);
      goBack();
    }
  };

  const deleteHabitAndClose = () => {
    deleteHabit(id);
    goBack();
  };

  return (
    <form
      onSubmit={saveAndClose}
      className={`mobile-size-container ${
        validityChecked ? "was-validated" : ""
      }`}
      noValidate
    >
      <h2>Nice, what will it be?</h2>
      <div className="mb-3">
        <label htmlFor="new-habit-name">Name</label>
        <input
          id="new-habit-name"
          className="form-control"
          value={name}
          onChange={e => setName(e.target.value)}
          required
        ></input>
        <div className="invalid-feedback">
          Hey! Trying to commit to "Blank"?
        </div>
      </div>
      <div className="mb-3">
        <label htmlFor="new-habit-type">Type</label>
        <select
          id="new-habit-type"
          className="form-control"
          value={type}
          onChange={e => setType(e.target.value)}
          onBlur={e => setType(e.target.value)}
          required
        >
          <option value=""></option>
          <option value="Health">Health</option>
          <option value="Social">Social</option>
          <option value="Career">Career</option>
          <option value="Hobbies">Hobbies</option>
        </select>
        <div className="invalid-feedback">Common... Help me help you!</div>
      </div>
      <div className="mb-3">
        <label htmlFor="new-habit-frequency">Frequency</label>
        <select
          id="new-habit-frequency"
          className="form-control"
          value={frequency}
          onChange={e => setFrequency(e.target.value)}
          onBlur={e => setFrequency(e.target.value)}
          required
        >
          <option value=""></option>
          <option value="1">1x</option>
          <option value="2">2x</option>
          <option value="3">3x</option>
          <option value="4">4x</option>
          <option value="5">5x</option>
          <option value="6">6x</option>
          <option value="7">7x</option>
        </select>
        <div className="invalid-feedback">
          Funny, that's an easy way to ace it!
        </div>
      </div>
      <hr className="mb-4" />
      <button className="btn btn-primary btn-lg btn-block">Let's start!</button>
      {id ? (
        <button
          type="button"
          className="btn btn-danger btn-lg btn-block"
          onClick={deleteHabitAndClose}
        >
          Let's scrap this!
        </button>
      ) : null}
      <button
        type="button"
        className="btn btn-secondary btn-lg btn-block"
        onClick={goBack}
      >
        Maybe tomorrow...
      </button>
    </form>
  );
};

export default SaveHabit;
