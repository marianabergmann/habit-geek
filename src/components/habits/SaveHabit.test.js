import React from "react";
import { shallow } from "enzyme";
import SaveHabit from "./SaveHabit";

let fakeEvent;
let saveHabitMock;
let deleteHabitMock;
let goBackMock;

describe("SaveHabit component", () => {
  beforeEach(() => {
    saveHabitMock = jest.fn();
    deleteHabitMock = jest.fn();
    goBackMock = jest.fn();
    fakeEvent = {
      preventDefault: jest.fn(),
      target: { checkValidity: jest.fn().mockReturnValue(true) }
    };
  });

  it("Renders static elements", () => {
    const saveHabits = shallow(
      <SaveHabit name="name" frequency="frequency" type="1"></SaveHabit>
    );
    expect(saveHabits).toMatchSnapshot();
  });

  it("creates a habit", () => {
    const saveHabits = shallow(
      <SaveHabit saveHabit={saveHabitMock} goBack={goBackMock}></SaveHabit>
    );
    saveHabits
      .find("#new-habit-name")
      .simulate("change", { target: { value: "Read" } });
    saveHabits
      .find("#new-habit-type")
      .simulate("change", { target: { value: "Social" } });
    saveHabits
      .find("#new-habit-frequency")
      .simulate("change", { target: { value: "1x" } });
    saveHabits.find("form").simulate("submit", fakeEvent);
    expect(saveHabitMock).toHaveBeenCalledWith(
      undefined,
      "Read",
      "Social",
      "1x"
    );
    expect(goBackMock).toBeCalled();
  });

  it("updates a habit", () => {
    const saveHabits = shallow(
      <SaveHabit
        id="read"
        saveHabit={saveHabitMock}
        goBack={goBackMock}
      ></SaveHabit>
    );
    saveHabits
      .find("#new-habit-name")
      .simulate("change", { target: { value: "Read" } });
    saveHabits
      .find("#new-habit-type")
      .simulate("change", { target: { value: "Social" } });
    saveHabits
      .find("#new-habit-frequency")
      .simulate("change", { target: { value: "1x" } });
    saveHabits.find("form").simulate("submit", fakeEvent);
    expect(saveHabitMock).toHaveBeenCalledWith("read", "Read", "Social", "1x");
    expect(goBackMock).toBeCalled();
  });

  it("validates forms", () => {
    const saveHabit = shallow(
      <SaveHabit saveHabit={saveHabitMock} goBack={goBackMock}></SaveHabit>
    );
    fakeEvent.target.checkValidity.mockReturnValueOnce(false);
    saveHabit.find("form").simulate("submit", fakeEvent);
    expect(saveHabit.find(".was-validated").length).toBe(1);
    expect(fakeEvent.preventDefault).toHaveBeenCalledTimes(1);
    expect(fakeEvent.target.checkValidity).toHaveBeenCalledTimes(1);
    expect(saveHabitMock).toHaveBeenCalledTimes(0);
    expect(goBackMock).toHaveBeenCalledTimes(0);
  });

  it("deletes a habit", () => {
    const saveHabits = shallow(
      <SaveHabit
        id="read"
        deleteHabit={deleteHabitMock}
        goBack={goBackMock}
      ></SaveHabit>
    );
    saveHabits.find(".btn-danger").simulate("click");
    expect(deleteHabitMock).toHaveBeenCalledWith("read");
    expect(goBackMock).toBeCalled();
  });

  it("selects change on blur for accessibility", () => {
    const saveHabits = shallow(
      <SaveHabit saveHabit={saveHabitMock} goBack={goBackMock}></SaveHabit>
    );
    saveHabits
      .find("#new-habit-type")
      .simulate("blur", { target: { value: "Social" } });
    saveHabits
      .find("#new-habit-frequency")
      .simulate("blur", { target: { value: "1x" } });
    saveHabits.find("form").simulate("submit", fakeEvent);
    expect(saveHabitMock).toHaveBeenCalledWith(undefined, "", "Social", "1x");
  });

  it("cancels habit creation", () => {
    const saveHabits = shallow(<SaveHabit goBack={goBackMock}></SaveHabit>);
    saveHabits.find(".btn-secondary").simulate("click");
    expect(goBackMock).toHaveBeenCalled();
  });
});
