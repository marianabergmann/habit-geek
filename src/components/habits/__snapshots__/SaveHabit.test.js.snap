// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveHabit component Renders static elements 1`] = `
<form
  className="mobile-size-container "
  noValidate={true}
  onSubmit={[Function]}
>
  <h2>
    Nice, what will it be?
  </h2>
  <div
    className="mb-3"
  >
    <label
      htmlFor="new-habit-name"
    >
      Name
    </label>
    <input
      className="form-control"
      id="new-habit-name"
      onChange={[Function]}
      required={true}
      value="name"
    />
    <div
      className="invalid-feedback"
    >
      Hey! Trying to commit to "Blank"?
    </div>
  </div>
  <div
    className="mb-3"
  >
    <label
      htmlFor="new-habit-type"
    >
      Type
    </label>
    <select
      className="form-control"
      id="new-habit-type"
      onBlur={[Function]}
      onChange={[Function]}
      required={true}
      value="1"
    >
      <option
        value=""
      />
      <option
        value="Health"
      >
        Health
      </option>
      <option
        value="Social"
      >
        Social
      </option>
      <option
        value="Career"
      >
        Career
      </option>
      <option
        value="Hobbies"
      >
        Hobbies
      </option>
    </select>
    <div
      className="invalid-feedback"
    >
      Common... Help me help you!
    </div>
  </div>
  <div
    className="mb-3"
  >
    <label
      htmlFor="new-habit-frequency"
    >
      Frequency
    </label>
    <select
      className="form-control"
      id="new-habit-frequency"
      onBlur={[Function]}
      onChange={[Function]}
      required={true}
      value="frequency"
    >
      <option
        value=""
      />
      <option
        value="1"
      >
        1x
      </option>
      <option
        value="2"
      >
        2x
      </option>
      <option
        value="3"
      >
        3x
      </option>
      <option
        value="4"
      >
        4x
      </option>
      <option
        value="5"
      >
        5x
      </option>
      <option
        value="6"
      >
        6x
      </option>
      <option
        value="7"
      >
        7x
      </option>
    </select>
    <div
      className="invalid-feedback"
    >
      Funny, that's an easy way to ace it!
    </div>
  </div>
  <hr
    className="mb-4"
  />
  <button
    className="btn btn-primary btn-lg btn-block"
    type="submit"
  >
    Let's start!
  </button>
  <button
    className="btn btn-secondary btn-lg btn-block"
  >
    Maybe tomorrow...
  </button>
</form>
`;
